### ISSUE TITLE
Concurrency issue while processing multiple markdown files in parallel

### ISSUE THREAD

**Issue by SelfRef** [markedjs/marked-gfm-heading-id#617]
Hi,

I've built a markdown processing logic for my own website to convert markdown files on server side (Noje.js) directly into form of HTML + headings. The problem I have after adding `gfmHeadingId()` is that I cannot get a deterministic response every time I process the files asynchronously and use `getHeadingList()`.

Often the headings I get are correctly assigned to files but once every few times I get none headings or all heading of all files from single call of `getHeadingList()`.

Function generating HTML:
```ts
export async function convertMdToHtml(rawMarkdown: string) {
	let headings;
	const marked = new Marked()
	marked.use(
		gfmHeadingId(),
		{
			hooks: {
				postprocess(html) {
					headings = getHeadingList()
					return html
				},
			}
		}
	)

	const rawHtml = await marked.parse(rawMarkdown)
	const window = new JSDOM().window;
	const purify = DOMPurify(window);
	const html = purify.sanitize(rawHtml)

	return {
		html,
		headings,
	}
}
```

Calling code:
```ts
const notePromisses = noteList.map(async note => {
	let fileContent = await storage.getItem(note);
	const data = await convertMdToHtml(fileContent)

	return {
		name: note,
		data,
	}
})

const notes = await Promise.all(notePromisses)
```

I can't think of a solution that would allow resolving this issue without getting rid of asynchronous calls, which on the other side is resulting in long response time, which is not optimal. For the moment I'm trying to work around that with not using `getHeadingList()` at all but instead by manually pulling a list of headings with id attribute added, so basically reimplementing the functionality of `getHeadingList()` as a separate subroutine.

**Comment by SelfRef** [comment]
If someone is goint to have the same issue, there's my simple solution (good enough for me at least):

```ts
const rawHtml = await marked.parse(markdown)
const headings: HeadingData[] = rawHtml.matchAll(/<h([1-6]).*id="(.+)".*>(.+)<\/h/g).map(m => ({
	level: Number.parseInt(m[1]),
	id: m[2],
	text: m[3],
	raw: m[3],
})).toArray()
```

**Comment by UziTech** [comment]
I can't seem to reproduce this. Javascript isn't multi threaded so there is no true concurrency (unless you are using workers) and since you are not setting `async: true` in marked `marked.parse` returns a string not a promise so it isn't running asynchronously even though you `await` it.

If you can create a repo with an example I can run to get the erroneous output I can help debug but I can't find anything obviously wrong with running it this way.

**Comment by SelfRef** [comment]
Thank you for the response. The missing `{async: true}` is my mistake during creating this example, the code I'm using does have async turn on and more marked extensions. I also forgot to mention the code I added is used in server part of Nuxt application, so in the Node environment.

I created a PoC with intention to recreate the issue here, but oddly enough I can't no matter how many times I refresh the page. It might mean the problem is something else, and it may be related with production environment where I saw the bug.

![image](https://github.com/user-attachments/assets/7d83b808-b995-4992-bce6-93ec44852d82)

Thank you for your time. I [leave my PoC](https://github.com/SelfRef/experiments/tree/master/markedjs-test), maybe it will be useful to someone.
